python/dgl/transform.py

```diff
@@ -743,38 +743,46 @@ def compact_graphs(graphs, always_preserve=None):
 
     return new_graphs
 
-def to_block(g, rhs_nodes=None, lhs_suffix="_l", rhs_suffix="_r"):
+def to_block(g, dst_nodes=None):
     """Convert a graph into a bipartite-structured "block" for message passing.
 
-    Specifically, we create one node type ``ntype_l`` on the "left hand" side and another
-    node type ``ntype_r`` on the "right hand" side for each node type ``ntype``.  The
-    nodes of type ``ntype_r`` would contain the nodes that have an inbound edge of any type,
-    while ``ntype_l`` would contain all the nodes on the right hand side, as well as any
-    nodes that have an outbound edge of any type pointing to any node on the right hand side.
+    A block graph is uni-directional bipartite graph consisting of two sets of nodes
+    SRC and DST. Each set can have many node types while all the edges are from SRC
+    nodes to DST nodes.
 
-    For each relation graph of canonical edge type ``(utype, etype, vtype)``, edges
-    from node type ``utype`` to node type ``vtype`` are preserved, except that the
-    source node type and destination node type become ``utype_l`` and ``vtype_r`` in
-    the new graph.  The resulting relation graph would have a canonical edge type
-    ``(utype_l, etype, vtype_r)``.
+    Specifically, for each relation graph of canonical edge type ``(utype, etype, vtype)``,
+    node type ``utype`` belongs to SRC while ``vtype`` belongs to DST.
+    Edges from node type ``utype`` to node type ``vtype`` are preserved. If
+    ``utype == vtype``, the result graph will have two node types of the same name ``utype``,
+    but one belongs to SRC while the other belongs to DST. This is because although
+    they have the same name, their node ids are relabeled differently (see below). In
+    both cases, the canonical edge type in the new graph is still
+    ``(utype, etype, vtype)``, so there is no difference when referring to it.
 
-    We refer to such bipartite-structured graphs a **block**.
+    Moreover, the function also relabels node ids in each type to make the graph more compact.
+    Specifically, the nodes of type ``vtype`` would contain the nodes that have at least one
+    inbound edge of any type, while ``utype`` would contain all the DST nodes of type ``utype``,
+    as well as the nodes that have at least one outbound edge to any DST node.
 
-    If ``rhs_nodes`` is given, the right hand side would contain the given nodes.
-    Otherwise, the right hand side would be determined by DGL via the rules above.
+    Since DST nodes are included in SRC nodes, a common requirement is to fetch
+    the DST node features from the SRC nodes features. To avoid expensive sparse lookup,
+    the function assures that the DST nodes in both SRC and DST sets have the same ids.
+    As a result, given the node feature tensor ``X`` of type ``utype``,
+    the following code finds the corresponding DST node features of type ``vtype``:
+
+    .. code::
+
+        X[:block.number_of_nodes('vtype')]
```

Comment:
If we are converting a block with `'user'` as `utype` and again `'user'` as `vtype` (i.e. I'm converting a homograph to a block), would this statement work?

Reply:
It should be `X[:block.number_of_nodes('DST/vtype')]`. Will change that.